File: leetcode/101-200/162-find-peak-element/bench/find_peak.c
#include <stdio.h>
#include <stdlib.h>

static long find_peak(const long *arr, long lo, long hi) {
    while (lo < hi) {
        long mid = lo + (hi - lo) / 2;
        if (arr[mid] < arr[mid + 1]) {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    return lo;
}

int main(void) {
    long n = 4000000, window = 4096, passes = 1000000;
    long *arr = malloc(n * sizeof(long));
    long state = 12345;
    for (long c = 0; c < n; c++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        arr[c] = state % 1000003;
    }
    long span = n - window;
    long sink = 0;
    for (long p = 0; p < passes; p++) {
        long base = (p * 4099L) % span;
        arr[base] = (arr[base] + 1) % 1000003;
        sink += find_peak(arr, base, base + window - 1);
    }
    printf("%ld\n", sink);
    free(arr);
    return 0;
}

File: leetcode/101-200/162-find-peak-element/bench/find_peak.rs
fn find_peak(arr: &[i64], lo0: i64, hi0: i64) -> i64 {
    let mut lo = lo0;
    let mut hi = hi0;
    while lo < hi {
        let mid = lo + (hi - lo) / 2;
        if arr[mid as usize] < arr[(mid + 1) as usize] {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    lo
}

fn main() {
    let n: i64 = 4000000;
    let window: i64 = 4096;
    let passes: i64 = 1000000;
    let mut arr = vec![0i64; n as usize];
    let mut state: i64 = 12345;
    for c in 0..n as usize {
        state = (state * 1103515245 + 12345) & 2147483647;
        arr[c] = state % 1000003;
    }
    let span = n - window;
    let mut sink: i64 = 0;
    for p in 0..passes {
        let base = (p * 4099) % span;
        let bu = base as usize;
        arr[bu] = (arr[bu] + 1) % 1000003;
        sink += find_peak(&arr, base, base + window - 1);
    }
    println!("{}", sink);
}
